Notebooks tab redesign

rtitle opened the pull request:
Relevant Leo ticket: https://github.com/DataBiosphere/leonardo/issues/281

This change set refactors the Notebooks tab in the following way:
- Instead of displaying a list of clusters, it now displays a list of notebooks
- Notebooks can be associated and opened with a cluster
- Notebooks are localized to the cluster and delocalized back to the workspace bucket
- Notebooks can be copied, renamed, deleted

It looks like this:

![screen shot 2018-08-03 at 10 53 48 am](https://user-images.githubusercontent.com/5368863/43649604-8ea9c430-970b-11e8-998f-60a1a0dfc345.png)

Lou helped with the design and UX a while back. It actually brings the experience a little more in line with Saturn (in fact, I borrowed some Saturn code). These changes are needed for some high-profile workshops where we'd like to show notebooks-in-the-workspace in FireCloud.

There are still a few TODOs:
- Upload notebook, cluster details is not fully complete -- noted in the code
- Clone workspace is not implemented, this is tracked more closely here: https://github.com/DataBiosphere/saturn-ui/issues/277
- No automation tests yet. We should have some level of automated testing, I'll probably work on those in parallel with the review.

But it works!! Please take a look and let me know of any feedback.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

rtitle committed:
Initial notebooks table

rtitle committed:
Localization stub

rtitle committed:
Progress

rtitle committed:
Cluster details, delete cluster

rtitle committed:
post rebase, resolve some todos

rtitle committed:
Rename/Duplicate/Delete notebooks

rtitle committed:
Notebook functionality and code cleanup

rtitle committed:
Persistence

rtitle committed:
Add placehold upload button

rtitle committed:
Self review feedback

rtitle commented on src/cljs/main/broadfcui/components/tab_bar.cljs:
```diff
@@ -11,8 +11,7 @@
    (fn [{:keys [props state]}]
      (let [{:keys [data-test-id first? active? label href on-refresh]} props]
        [:a {:data-test-id data-test-id
-            ;; Notebooks tab label has different padding because of the temporary "BETA" tag
```
Removed BETA banner

rtitle commented on src/cljs/main/broadfcui/page/workspace/notebooks/cluster_details.cljs:
```diff
@@ -0,0 +1,44 @@
+          [:div {}
+           [comps/ErrorViewer {:error server-error}]
+           (if cluster-details
+             (str cluster-details) ;TODO display in a nicer way than just raw json
```
This is still a TODO -- I'd like to display cluster details in some sort of a tabular way.

zarsky-broad replied:
```cljs
(js/JSON.stringify cluster-details nil 2)
```

rtitle commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_cluster.cljs:
```diff
@@ -0,0 +1,208 @@
+                   "n1-highmem-64"
+                   "n1-highmem-96"])
+
+(react/defc ClusterCreator
```
This code is essentially copied from here: https://github.com/broadinstitute/firecloud-ui/blob/develop/src/cljs/main/broadfcui/page/workspace/notebooks/notebooks.cljs#L82

rtitle commented on src/cljs/main/broadfcui/page/workspace/notebooks/notebooks_table.cljs:
```diff
@@ -0,0 +1,443 @@
+(ns broadfcui.page.workspace.notebooks.notebooks_table
```
This file is where the dragons are :)

rtitle commented on src/cljs/main/broadfcui/page/workspace/notebooks/notebooks_table.cljs:
```diff
@@ -0,0 +1,443 @@
+             :toolbar-items [flex/spring [buttons/Button {:data-test-id "create-modal-button"
+                                                          :text "Create Notebook..." :style {:marginRight 1}
+                                                          :onClick #(swap! state assoc :show-create-dialog? true)}]
+                             ; TODO: upload notebook not yet implemented
```
Noting another TODO -- I realized late that we'd need "upload notebook" in addition to "create notebook".

rtitle replied:
This is implemented now

rtitle commented on src/cljs/main/broadfcui/page/workspace/notebooks/notebooks_table.cljs:
```diff
@@ -0,0 +1,443 @@
+     (swap! locals assoc :dead? true)
+     (.removeEventListener js/window "message" (react/method this :-notebook-extension-listener)))
+
+   ; Communicates with the Leo notebook extension.
```
A lot of this code is copied from here: https://github.com/broadinstitute/firecloud-ui/blob/develop/src/cljs/main/broadfcui/page/workspace/notebooks/notebooks.cljs#L458

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/choose_cluster.cljs:
```diff
@@ -0,0 +1,47 @@
+(ns broadfcui.page.workspace.notebooks.choose_cluster
+  (:require
+   [dmohs.react :as react]
+   [broadfcui.components.modals :as modals]
```
Nit: alphabetize. See [CONTRIBUTING.md](https://github.com/broadinstitute/firecloud-ui/blob/develop/CONTRIBUTING.md#requiring-namespaces)

zarsky-broad commented:
@rtitle Could you make sure you have the cljs settings imported in Intellij and run Code -> Reformat Code on the files you touched? (See [CONTRIBUTING.md](https://github.com/broadinstitute/firecloud-ui/blob/develop/CONTRIBUTING.md#styling-in-editors))

rtitle commented:
Will do

rtitle committed:
Reformat code

rtitle commented:
@zarsky-broad I imported the settings and reformatted all the notebook files. It definitely caught some stuff.

rtitle committed:
Add notebook upload button

rtitle committed:
A few more UX improvements:
- if there are no clusters, only the option to create a cluster is presented (not choose)
- error when uploading a notebook with the same name as an existing one

rtitle committed:
Fix upload logic error

rtitle committed:
Fix double .ipynb during upload

rtitle committed:
Just 1 more: fix extra json escaping of notebook uploads

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/choose_cluster.cljs:
```diff
@@ -0,0 +1,47 @@
+   (fn [{:keys [props state this]}]
+     (let [{:keys [cluster-map choose-notebook]} props
+           cluster-name (:cluster-name choose-notebook)
+           cluster-menu-entry (fn [c] (str (:clusterName c) " (" (:status c) ")"))]
```
I might prefer to destructure `c` so this function is:
```clojure
(fn [{:keys [clusterName status]}] (str clusterName " (" status ")"))
```

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/choose_cluster.cljs:
```diff
@@ -0,0 +1,47 @@
+   :-create-cluster
+   (fn [{:keys [props refs]}]
+     ((:show-create-cluster props))
+     ((:dismiss props)))})
```
There's a common opinion that the double parens is ugly, so it might be better to pull these out of props first:
```clojure
(let [{:keys [show-create-cluster dismiss]} props]
  (show-create-cluster)
  (dismiss))
```
Same in `:-choose-cluster` above, and in several other places throughout.

zarsky-broad replied:
Also, I believe we standardized on ending files with a newline.

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_cluster.cljs:
```diff
@@ -0,0 +1,208 @@
+                                                }
+                                        :href "javascript:;"
+                                        :onClick (fn [] (swap! state #(-> % (assoc :label-gensym (gensym))
+                                                                          (update :labels utils/delete i))))}
```
I wrote `utils/multi-swap!` to handle exactly this kind of thing.

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_cluster.cljs:
```diff
@@ -0,0 +1,208 @@
+                                 (:labels @state))])
+                 [buttons/Button {:text "Add Label" :icon :add-new :style {:marginBottom 10} :data-test-id "add-label-button"
+                                  :onClick (fn []
+                                             (swap! state #(-> %
```
multi-swap!

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_notebook.cljs:
```diff
@@ -0,0 +1,115 @@
+       (if fails
+         (swap! state assoc :validation-errors fails)
+         ; fail if a notebook already exists with the same name
+         (if (some (comp (partial = new-notebook-name) #(notebook-utils/notebook-name %)) notebooks)
```
`#(notebook-utils/notebook-name %)` => `notebook-utils/notebook-name`

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/duplicate_notebook.cljs:
```diff
@@ -0,0 +1,56 @@
+            [FoundationTooltip {:text (notebook-utils/create-inline-form-label "Name")
+                                :tooltip (str "Enter new name for copy of notebook \"" (notebook-utils/notebook-name choose-notebook) \" "")}]]
+           [input/TextField {:data-test-id "notebook-name-input" :ref "newNotebookName" :autoFocus true :style {:width "100%"}
+                             :defaultValue (str "Copy of " (notebook-utils/notebook-name choose-notebook)) :predicates [(input/nonempty "Notebook name") (input/alphanumeric_-space "Notebook name")]}]
```
`:predicates` on a new line for clarity

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/duplicate_notebook.cljs:
```diff
@@ -0,0 +1,56 @@
+         (swap! state assoc :validation-errors fails)
+         (if (= (notebook-utils/notebook-name choose-notebook) new-notebook-name)
+           ((:dismiss props)) ; no-op if the name is unchanged
+           (if (some (comp (partial = new-notebook-name) #(notebook-utils/notebook-name %)) notebooks)
```
same as before, and throughout your changes.

MattPutnam commented on src/cljs/main/broadfcui/page/workspace/notebooks/notebooks_table.cljs:
```diff
@@ -0,0 +1,512 @@
+     (let [{:keys [choose-notebook]} @state
+           {:keys [notebooks cluster-map pet-token toolbar-items]} props]
+       [:div {}
+        (when (:show-cluster-choose-dialog? @state)
```
There's a utility `broadfcui.components.modals/show-modals` for handling large numbers of modals like this. It also sticks the `:dismiss` on there for you.

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/choose_cluster.cljs:
```diff
@@ -0,0 +1,47 @@
+           cluster-menu-entry (fn [c] (str (:clusterName c) " (" (:status c) ")"))]
+       [modals/OKCancelForm
+        {:header "Choose Cluster"
+         :dismiss (:dismiss props)
```
Why not also destructure `dismiss` from `props` where you do the rest above?

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/choose_cluster.cljs:
```diff
@@ -0,0 +1,47 @@
+         :content
+         (react/create-element
+          [:div {:style {:marginBottom -20}}
+           (notebook-utils/create-inline-form-label (str "Choose a cluster to associate with notebook \""
```
Consider putting the argument to long-names methods like this on the next line, for readability.

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/cluster_details.cljs:
```diff
@@ -0,0 +1,44 @@
+   :-get-cluster-details
+   (fn [{:keys [state this props]}]
+     (let [{:keys [cluster-to-display-details]} props]
+       (swap! state assoc :querying? true)
```
`utils/multi-swap!`

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/cluster_error.cljs:
```diff
@@ -0,0 +1,46 @@
+           {:keys [cluster-to-view]} props]
+       [modals/OKCancelForm
+        {:header "Cluster Error"
+         :dismiss (:dismiss props)
```
As above, why not destructure

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_cluster.cljs:
```diff
@@ -0,0 +1,208 @@
+                                                                          (update :labels utils/delete i))))}
+                                       (icons/render-icon {:style {:marginTop "35%"}} :remove))
+                                     [input/TextField {:data-test-id (str "key-" i "-input")
+                                                       :style {:ref (str "key" i) :marginBottom 0 :width "48%" :marginRight "4%"}
```
`:ref` should (probably) not be inside `:style`

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_cluster.cljs:
```diff
@@ -0,0 +1,208 @@
+                  [FoundationTooltip {:text (style/create-form-label "Custom Script URI")
+                                      :tooltip "The GCS URI of a bash script you wish to run on your cluster before it starts up."}]]
+                 [input/TextField {:data-test-id "custom-script-uri-input" :ref "userScriptURI" :autoFocus true :style {:width "100%"}}]
+                 (when (seq (:labels @state))
```
A good rule of thumb is that, unless you're doing something funky, your only reference to `@state` should be in the `let` at the top of a function, and same from `props`. That way, it's easy to see at a glance what properties are used in a function, in case it needs to be refactored or replaced.

zarsky-broad commented on src/cljs/main/broadfcui/page/workspace/notebooks/create_notebook.cljs:
```diff
@@ -0,0 +1,115 @@
+(ns broadfcui.page.workspace.notebooks.create_notebook
```
We usually name _namespaces_ with a dash and _file names_ with an underscore

zarsky-broad reviewed (approved):
Not my most thorough review, and didn't run, but basically looks ok. Some comments, mostly stylistic.

rtitle commented:
Thanks for the reviews @zarsky-broad and @MattPutnam! I'll look at making those changes.

rtitle committed:
PR feedback

missed a destructure

rtitle commented:
@zarsky-broad @MattPutnam I applied PR feedback (see latest [commit](https://github.com/broadinstitute/firecloud-ui/pull/1378/commits/7073bc6aad23ecd30ab598509a1ecb3c735b764a)) and rebased from develop, in case you want to take another look. I'm still finalizing doc updates with Comms, but I'm hoping to get this merged fairly soon.

rtitle committed:
Add back beta

rtitle committed:
Cluster details page

rtitle committed:
Use createCluster v2 endpoint
